webgl: Add basic support for framebuffer attachments

anholt opened the pull request:
This is by no means a complete implementation, but I've slowed down on working on it, so I think we should look at what it takes to merge the current code.  There are some major features missing, like initializing renderbuffers to 0 (uninitialized memory leak), tracking the attachments' attributes (width/height/format) for parameter requests, and lots of missing glCheckFramebufferStatus() validation.  On the other hand, this is enough to run some demos using FBOs.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13639 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13872)

<!-- Reviewable:end -->

anholt committed:
webgl: Add support for renderbufferStorage().

This is not a complete implementation yet: It doesn't clear the
contents of the renderbuffer on creation.  However, Gecko's plan to
only clear renderbuffers when the first FBO using them is the
simplest.

anholt committed:
webgl: Add support for checkFramebufferStatus().

For now it's returning the default UNSUPPORTED on user FBOs.

object-deletion-behaviour.html starts running a bunch more subtets.

anholt committed:
webgl: Add support for FBO attachments.

This allows many FBO tests to start running as their framebuffers
start coming back as framebuffer complete.

anholt committed:
webgl: Detach RBs and textures from the bound FBO on deletion.

This is part of general GL behavior: when an object is deleted, look
through the currently bound objects and detach the deleted object from
them.  Detaching an object from an FBO causes it to need to be
re-checked for its status.

anholt committed:
webgl: Re-check FBO status on binding an FBO.

When the FBO has been unbound, anything that happened to its
attachments (deletion, new binding, reallocation) may have caused it
to change status.

This doesn't change any test results currently, because we're not
checking if the attachments are deleted, the wrong size, or have
appropriate formats.

anholt committed:
webgl: Throw an error when binding a deleted framebuffer.

The spec was recently changed to clarify that this should throw an
error.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webglframebuffer.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGLRenderingContext.webidl, components/script/dom/webglrenderbuffer.rs
 * @fitzgen: components/script/dom/webglframebuffer.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGLRenderingContext.webidl, components/script/dom/webglrenderbuffer.rs
 * @emilio: components/script/dom/webglframebuffer.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webglrenderbuffer.rs

glennw commented:
r? @emilio

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -806,7 +806,11 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
         self.bound_framebuffer.set(framebuffer);
         if let Some(framebuffer) = framebuffer {
-            framebuffer.bind(target)
+            if framebuffer.is_deleted() {
```
The spec changed and this case should be ignored now (don't change the state of the binding), plus produce an `INVALID_OP`.

emilio commented on components/script/dom/webglrenderbuffer.rs:
```diff
@@ -33,6 +35,7 @@ impl WebGLRenderbuffer {
             ever_bound: Cell::new(false),
             is_deleted: Cell::new(false),
             renderer: renderer,
+            internalformat: Cell::new(None),
```
I think the usual way to name members is using snake case (i.e. `internal_format`). Not a huge deal though.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2536,6 +2536,39 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+        // and we have to throw out negative-size values as well just
+        // like for TexImage.
+        //
+        // FIXME: Handle max_renderbuffer_size, which doesn't seem to
```
Yeah, these are being added on demand, I'll file an issue about this.

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -80,10 +105,139 @@ impl WebGLFramebuffer {
+                let is_cube = match textarget {
+                    constants::TEXTURE_2D => false,
+
+                    constants::TEXTURE_CUBE_MAP_POSITIVE_Z => true,
```
Shouldn't you test the rest cube texture names?

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -240,6 +241,49 @@ impl WebGLFramebuffer {
     }
 
+    pub fn detach_renderbuffer(&self, rb: &WebGLRenderbuffer) {
+        let attachments = vec![&self.color,
```
I think this can be in the stack instead being a vector.

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -240,6 +241,49 @@ impl WebGLFramebuffer {
+    }
+
+    pub fn detach_texture(&self, texture: &WebGLTexture) {
+        let attachments = vec![&self.color,
```
ditto

emilio reviewed (changes requested):
Looks good, thanks for doing this @anholt! I have a few comments.

emilio commented:
Yay, let's land this. Thanks for doing this once again :)

@bors-servo: r+

bors-servo commented:
:pushpin: Commit 4134c1a has been approved by `emilio`

<!-- @bors-servo r=emilio 4134c1a19bc835166b5976fedb98819c872d38c6 -->

bors-servo commented:
:hourglass: Testing commit 4134c1a with merge a3c80e9...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/425)

jdm commented:
```
  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #278: at (16, 16) expected: 0,0,0,0 was 9,0,0,0
  │   → assert_true: at (16, 16) expected: 0,0,0,0 was 9,0,0,0 expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:153:5
  │ checkCanvasRectColor@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1047:9
  │ checkCanvasRect@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1073:3
  │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js line 1349 > eval:1:1
  │ shouldGenerateGLError@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1349:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:387:3
```

anholt commented:
I made a mistake before with my `handle_object_deletion` macro -- we need to clear the GL's binding, not just WebGL's, which should fix the test's problem.  However, this also made me notice #13901, and fixing that would at least make this failure not intermittent.

I won't have time for this tonight, though.

anholt commented:
OK, that failure does seem fairly stable, so I've added it and hopefully that will be enough.  There were two more subfailures in the same test that were showing up, too.  Not sure why I didn't get them last night.

emilio commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 2035ed9 has been approved by `emilio`

<!-- @bors-servo r=emilio 2035ed92f54d58a5cc8360695b5f83becb1d0cbe -->

bors-servo commented:
:hourglass: Testing commit 2035ed9 with merge cb5b9da...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/448)

jdm commented:
```
  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #281: at (0, 0) expected: 255,0,0,255 was 0,172,146,24
  │   → assert_true: at (0, 0) expected: 255,0,0,255 was 0,172,146,24 expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:153:5
  │ checkCanvasRectColor@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1047:9
  │ checkCanvasRect@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1073:3
  │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js line 1349 > eval:1:1
  │ shouldGenerateGLError@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1349:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:391:3
```

anholt committed:
webgl: Unbind deleted objects from the GL side, not just our side.

Once FBOs are allowed, we were running into testcases that deleted the
active FBO, and expected the following ReadPixels to come from the
default framebuffer.

anholt commented:
OK: Expecting the winsys backbuffer's red after glDeleteFramebuffer() of our FBO: that should be fixed in the new unbind commit.  Previously we were returning undefined.

emilio commented:
Sounds good.

@bors-servo: r+

bors-servo commented:
:pushpin: Commit 71d2660 has been approved by `emilio`

<!-- @bors-servo r=emilio 71d266052b09c031707139466163a7087f0acc19 -->

bors-servo commented:
:hourglass: Testing commit 71d2660 with merge e572045...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3789)

jdm commented:
```
LLVM ERROR: IO failure on output stream.
error: Could not compile `embedding`.
```
That's akward.

jdm commented:
@bors-servo: retry
Maybe it's not a permanent failure?

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3714), [arm64](http://build.servo.org/builders/arm64/builds/3725), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/434), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/431), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3793), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3781), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/473), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/475), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1694) are reusable. Rebuilding only [linux-dev](http://build.servo.org/builders/linux-dev/builds/3789)...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3792)

emilio commented:
@bors-servo: retry

 * Builder is out of disk space

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3714), [arm64](http://build.servo.org/builders/arm64/builds/3725), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/434), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/431), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3793), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3781), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/473), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/475), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1694) are reusable. Rebuilding only [linux-dev](http://build.servo.org/builders/linux-dev/builds/3792)...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3795)

larsbergstrom commented:
@bors-servo retry
- servo-linux2 out of disk space

bors-servo commented:
:hourglass: Testing commit 71d2660 with merge dd41239...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/444)

KiChjang commented:
@bors-servo retry
* Disk space issues?

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3723), [arm64](http://build.servo.org/builders/arm64/builds/3734), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3805), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/443), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3803), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3790), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/483), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/484), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1706) are reusable. Rebuilding only [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/444)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/450)

larsbergstrom commented:
@bors-servo retry
- disk space issues on servo-linux2

bors-servo commented:
:hourglass: Testing commit 71d2660 with merge 8dd7ca6...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/460)

highfive commented:
▶ FAIL [expected PASS] /css21_dev/html4/abs-pos-non-replaced-vrl-136.htm
      └   → /css21_dev/html4/abs-pos-non-replaced-vrl-136.htm 3eb0ba81d5108e1d42ff482ad8689ea624b567d3
    /css21_dev/html4/reference/abs-pos-non-replaced-vrl-004-ref.htm 82d6b4e7e7a391559f9819448a87076b00477338
    Testing 3eb0ba81d5108e1d42ff482ad8689ea624b567d3 == 82d6b4e7e7a391559f9819448a87076b00477338

jdm commented:
@bors-servo: retry
* #13919

bors-servo commented:
:hourglass: Testing commit 71d2660 with merge fbec79e...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3743), [arm64](http://build.servo.org/builders/arm64/builds/3754), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3828), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/469), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/466), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3823), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3811), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/503), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/507), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1729)
